Field Arguments

jamesdh opened the pull request:
Fixes #53. Sorry for taking so long to submit this!

A core part of the GraphQL spec is the [ability for fields to take arguments](https://spec.graphql.org/July2015/#sec-Field-Arguments). This PR adds the ability to define custom fields on an object which in turn take arguments. It does not contain the ability to define arguments on existing fields. 

I may have been overzealous in upgrading some of the gradle bits. Feel free to let me know if you'd like me to resubmit without those changes.  

This PR does not yet contain any related documentation changes, hence the draft status. Wanted to get a little feedback before putting any time into them.

jamesdh committed:
Upgrade gradle and use latest common build scripts

jamesdh committed:
build-scan plugin no longer works w/ gradle 6+

jamesdh committed:
Promote for reuse by fields

jamesdh committed:
This has been resolved in groovy 2.5.6+

jamesdh committed:
Explicitly upcast

Appears to be some quirk with Traits perhaps, as this should implicitly upcast into an Object but it fails within this trait. This works everywhere else the ExecutesClosures trait is used.

jamesdh committed:
Extract arguments into a trait

Lets us be a bit more DRY as we reuse this

jamesdh committed:
Fixes static compilation issue

jamesdh committed:
Use arguments on properties

With the exception of PersistentGraphQLProperties

jamesdh committed:
Support building field-level arguments

jamesdh committed:
Fix static compilation issues when defining custom data fetchers

jamesdh committed:
Tests for field-level arguments

jamesdh commented on core/src/main/groovy/org/grails/gorm/graphql/types/output/AbstractObjectTypeBuilder.groovy:
```diff
@@ -50,12 +52,17 @@ abstract class AbstractObjectTypeBuilder implements ObjectTypeBuilder {
                 .deprecate(prop.deprecationReason)
                 .description(prop.description)
 
+        if(prop instanceof CustomGraphQLProperty) {
```
This felt a little gross to me, but no other place stood out as a better alternative for actually building the property-level arguments.

jameskleeh replied:
Agreed. You should use something like:

```
        if(prop instanceof Arguable) {
            List<GraphQLArgument> arguments = prop.getArguments(typeManager, mapping)
            if (!arguments.isEmpty()) {
                field.arguments(arguments)
            }
        }
```

jamesdh commented on core/src/main/groovy/org/grails/gorm/graphql/entity/operations/CustomOperation.groovy:
```diff
@@ -35,7 +35,7 @@ abstract class CustomOperation<T> implements Named<T>, Describable<T>, Deprecata
     boolean defaultListArguments = false
 
-    T dataFetcher(DataFetcher dataFetcher) {
+    T dataFetcher(DataFetcher<?> dataFetcher) {
```
This was a bit bizarre, but when the implementing GORM entity used @GrailsCompileStatic, this would result in a static compilation error without the generic wildcard.

jamesdh commented on core/src/main/groovy/org/grails/gorm/graphql/entity/dsl/helpers/Arguable.groovy:
```diff
@@ -46,7 +46,7 @@ trait Arguable<T> extends ExecutesClosures {
      * @return The operation in order to chain method calls
      */
-    T argument(String name, Class type, @DelegatesTo(value = SimpleArgument, strategy = Closure.DELEGATE_ONLY) Closure closure = null) {
+    T argument(String name, Class<?> type, @DelegatesTo(value = SimpleArgument, strategy = Closure.DELEGATE_ONLY) Closure closure = null) {
```
Similar, in that when the implementing GORM entity used @GrailsCompileStatic, this would result in a static compilation error without the generic wildcard.

jameskleeh reviewed (changes requested):
Overall this PR is pretty good. I think if we're going to introduce breaking changes by moving the argument classes, and modifying this method signature (https://github.com/grails/gorm-graphql/pull/56/files#diff-65191fd62016a5999609edc95174e48c4937c21dfe6bae9773874e3efbea5169R55), we should also do the upgrades to Groovy 3, etc and include it in Grails 4.1

Let me know if you would prefer to proceed without breaking changes and include this in an earlier release

jamesdh commented:
I'm fine with including it in Grails 4.1, although I'm a little unclear how it's a breaking change. None of the original functionality of the DSL is altered. 

I'll go ahead and write up some doc changes to go with it.

jamesdh committed:
Build on Arguable

See https://github.com/grails/gorm-graphql/pull/56#discussion_r527884980

jamesdh committed:
Arguments docs

- Pushed them up a level to help imply they can be used for both operations and fields
- Added a bit to the examples to emphasize usage

jamesdh commented:
@jameskleeh can you clarify what you meant by breaking changes? From what I can tell, nothing exposed in the public API/DSL has been altered, only additions. 

For the sake of just taking baby steps, I'd rather avoid doing the upgrade to Groovy 3/Grails 4.1 in this PR. I'd be happy to do so though in another though.

jameskleeh commented:
@jamesdh This is the breaking change https://github.com/grails/gorm-graphql/pull/56/files#diff-65191fd62016a5999609edc95174e48c4937c21dfe6bae9773874e3efbea5169R57. Someone could be extending this class

jamesdh committed:
Keep original method signature

And build field arguments as seperate method to avoid any possible breakages

jamesdh commented:
@jameskleeh this should be good now. Feel free to take another look!

jamesdh committed:
Fix codenarc violations

jamesdh commented:
@jeffbrown @jameskleeh @puneetbehl I understand you're all fairly focused on Micronaut these days and that's completely fair. But would it be possible to promote some additional people to this repo to move things along? There are lots of things that could be improved and I'd be happy to contribute some of my own time to do so, but at this point it's starting to feel all for naught.

jameskleeh commented:
@jamesdh Reviewing these PRs and creating a release is on my list of things to do. I assure you it is not all for naught

jameskleeh commented on core/src/main/groovy/org/grails/gorm/graphql/types/output/AbstractObjectTypeBuilder.groovy:
```diff
@@ -96,7 +99,11 @@ abstract class AbstractObjectTypeBuilder implements ObjectTypeBuilder {
                 if (prop.output) {
-                    fields.add(buildField(prop, NAME, entity.mappingContext).build())
+                    GraphQLFieldDefinition.Builder field = buildField(prop, NAME)
+                    if (prop instanceof Arguable) {
```
This if can be removed since its duplicated in the addFieldArgs method

jamesdh committed:
Duplicate check

jamesdh commented:
Thanks for the update and the review @jameskleeh!
